rfd/0138-postgres-backend.md

```diff
@@ -0,0 +1,132 @@
+
+Each auth server running the backend will periodically attempt to delete expired items - at an interval tentatively set to 30 seconds - deleting small batches of expired items in different transactions until no more items can be deleted.
+
+To generate an event stream we'll connect to the database with a regular connection, create a temporary logical replication slot, then poll changes on it via the [`pg_logical_slot_get_changes`](https://www.postgresql.org/docs/current/functions-admin.html#FUNCTIONS-REPLICATION) function. As the builtin `pgoutput` logical decoding plugin cannot be used without errors via the SQL interface except in very recent versions of Postgres (11.20, 12.15, 13.11, 14.8 or 15.3, released in May 2023), we use the third-party but very popular [`wal2json`](https://github.com/eulerto/wal2json) plugin, which is available by default on the managed Postgres offerings in Azure, AWS and GCP, and is installable from the Debian/Ubuntu official packages or from the Postgres APT and YUM repos.
+
+The `pg_logical_slot_get_changes` function (with the appropriate `wal2json` options) will return one entry per change, which maps nicely to the backend event model; updates and inserts will be rendered as `OpPut` events, deletes as `OpDelete` events.
+
+### General architecture of the audit log
+
+The database schema for the audit log will be composed of just an `events` table with the appropriate indexes for searching:
+
+```pgsql
+CREATE TABLE events (
+    event_time timestamptz NOT NULL,
+    event_id uuid NOT NULL,
+    event_type text NOT NULL,
+    session_id uuid NOT NULL,
+    event_data json NOT NULL,
+    creation_time timestamptz NOT NULL DEFAULT now(),
+    PRIMARY KEY (event_time, event_id)
+);
+CREATE INDEX events_creation_time_idx ON events (creation_time);
+CREATE INDEX events_search_session_events_idx ON events (session_id, event_time, event_id) WHERE session_id != '00000000-0000-0000-0000-000000000000'
+```
+
+The database schema is made to be compatible with the one for the backend, although we are going to recommend storing the backend and the audit log in two different databases (having both on the same database cluster is going to be fine), because logical decoding is applied per-database, and keeping both in the same database will require the logical decoding plugin to do extra work to filter out the changes to the audit log.
```

Comment:
Does it mean that we still need to take similar precautions about transaction sizes as above, or are audit log transactions guaranteed to be small enough?

Reply:
That's a very good point: logical decoding on a big database is going to take a long time even if the transaction doesn't involve tables that are filtered out; I think we're going to strongly discourage users from using the same database for both the backend and the audit log, and not accomodate sharing a database at all. I'll update the wording to reflect this.